File: 搜索与回溯/offer-32-3.py
import collections
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        res = []
        deque = collections.deque([root])
        cnt = 1
        while(deque):
            temp = collections.deque()
            for _ in range(len(deque)):
                node = deque.popleft()
                if cnt%2==0:
                    temp.append(node.val)
                else:
                    temp.appendleft(node.val)
                if node.left:
                    deque.append(node.left)
                if node.right:
                    deque.append(node.right)
            res.append(list(temp))
        return res


root = TreeNode(3)
root.left = TreeNode(9)
root.right = TreeNode(20)
root.right.left = TreeNode(15)
root.right.right = TreeNode(7)

obj = Solution()
print(obj.levelOrder(root))

File: 搜索与回溯/offer-32-3.cpp
#include<vector>
#include<deque>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(!root)
            return {{}};
        vector<vector<int>>result;
        deque<TreeNode*>queue;
        queue.push_back(root);
        bool flag=true;
        while(!queue.empty()){
            vector<int>tmp;
            int length = queue.size();
            for(int i=0;i<length;i++){
                TreeNode* node = queue.front();
                queue.pop_front();
                tmp.push_back(node->val);
                if(node->left)
                    queue.push_back(node->left);
                if(node->right)
                    queue.push_back(node->right);
            }
            if(flag)
                result.push_back(tmp);
            else{
                reverse(tmp.begin(),tmp.end());
                result.push_back(tmp);
            }
            flag = !flag;
        }
        return result;
    }
};